xml-trim/doc/xml-trim.1
.\" Automatically generated by Pandoc 2.3.1
.\"
.TH "xml\-trim" "1" "2019\-03\-21" "" "xml\-utils"
.hy
.SH NAME
.PP
xml\-trim \- Trim whitespace in XML elements
.SH SYNOPSIS
.IP
.nf
\f[C]
xml\-trim\ [\-N\ <ns=URL>]\ [\-nh?]\ <elem>...\ <\ <src>\ >\ <dst>
\f[]
.fi
.SH DESCRIPTION
.PP
The \f[I]xml\-trim\f[] utility trims whitespace around the text contents
of specified elements.
.PP
Whitespace is trimmed according to the following rules:
.IP "1." 3
Whitespace characters at the beginning of the first text node child of
the specified elements are removed.
.IP "2." 3
Whitespace characters at the end of the last text node child of the
specified elements are removed.
.IP "3." 3
If the \-n option is given, sequences of whitespace characters in all
text node children of the specified elements are converted to a single
space.
.SH OPTIONS
.TP
.B \-h \-?
Show usage message.
.RS
.RE
.TP
.B \-N <ns=URL>
Registers an XML namespace handle for URL, which can then be used when
specifying element names as options.
Multiple namespaces can be registered by specifying this option multiple
times.
.RS
.RE
.TP
.B \-n
Normalize space in the specified elements in addition to trimming
whitespace.
.RS
.RE
.TP
.B \-\-version
Show version information.
.RS
.RE
.TP
.B <elem>...
Elements to trim space on.
May include a namespace prefix if the namespace was registered with \-N.
<elem> may be either a simple element name (e.g., "para") which matches
all elements with the same name at any position, or an XPath expression
(e.g., "//section/para") for finer control.
.RS
.RE
.TP
.B <src>
The source XML file containing the elements to trim.
.RS
.RE
.TP
.B <dst>
The output file.
.RS
.RE
.SH EXAMPLES
.SS Without namespace
.IP
.nf
\f[C]
<section>
\ \ <para>
\ \ \ \ Hello\ world.
\ \ </para>
</section>
\f[]
.fi
.IP
.nf
\f[C]
$\ xml\-trim\ para\ <\ example.xml
\f[]
.fi
.IP
.nf
\f[C]
<section>
\ \ <para>Hello\ world.</para>
</section>
\f[]
.fi
.SS With namespace
.IP
.nf
\f[C]
<d:section>
\ \ <d:para>
\ \ \ \ Hello\ world.
\ \ </d:para>
</d:section>
\f[]
.fi
.IP
.nf
\f[C]
$\ xml\-trim\ \-N\ d=http://docbook.org/ns/docbook\ d:para\ <\ example.xml
\f[]
.fi
.IP
.nf
\f[C]
<d:section>
\ \ <d:para>Hello\ world.</d:para>
</d:section>
\f[]
.fi
.SS Normalizing space
.IP
.nf
\f[C]
<section>
\ \ <para>
\ \ \ \ This\ is\ a\ <emphasis>long</emphasis>
\ \ \ \ paragraph\ with\ both\ extra\ whitespace
\ \ \ \ before\ and\ after\ the\ text,\ and\ line
\ \ \ \ breaks\ entered\ by\ the\ author\ to\ wrap
\ \ \ \ the\ text\ on\ a\ certain\ column.
\ \ </para>
</section>
\f[]
.fi
.IP
.nf
\f[C]
$\ xml\-trim\ \-n\ para\ <\ example.xml
\f[]
.fi
.IP
.nf
\f[C]
<section>
\ \ <para>This\ is\ a\ <emphasis>long</emphasis>\ paragraph\ with\ both\ extra\ whitespace\ before
and\ after\ the\ text,\ and\ line\ breaks\ entered\ by\ the\ author\ to\ wrap\ the
text\ on\ a\ certain\ column.</para>
</section>
\f[]
.fi
.SH AUTHORS
khzae.net.


utils/xml-trim/xml-trim.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <getopt.h>
#include <stdbool.h>
#include <libxml/tree.h>
#include <libxml/xpath.h>
#include <libxml/xpathInternals.h>
#include "xml-utils.h"

#define PROG_NAME "xml-trim"
#define VERSION "3.2.1"

/* Remove whitespace on left end of string. */
char *strltrm(char *dst, const char *src)
{
	int start;
	for (start = 0; isspace(src[start]); ++start);
	sprintf(dst, "%s", src + start);
	return dst;
}

/* Remove whitespace on right end of string. */
char *strrtrm(char *dst, const char *src)
{
	int len, end;
	len = strlen(src);
	for (end = len - 1; isspace(src[end]); --end);
	sprintf(dst, "%.*s", end + 1, src);
	return dst;
}

/* Normalize space by replacing all sequences of whitespace characters with a
 * single space.
 */
char *strnorm(char *dst, const char *src)
{
	int i, j;
	j = 0;
	for (i = 0; src[i]; ++i) {
		if (isspace(src[i])) {
			dst[j] = ' ';
			while (isspace(src[i + 1])) {
				++i;
			}
		} else {
			dst[j] = src[i];
		}
		++j;
	}
	return dst;
}

/* Register an XML namespace with the XPath context. */
void register_ns(xmlXPathContextPtr ctx, char *optarg)
{
	char *prefix, *uri;

	prefix = strtok(optarg, "=");
	uri    = strtok(NULL, "");

	xmlXPathRegisterNs(ctx, BAD_CAST prefix, BAD_CAST uri);
}

/* Trim space in a text node. */
void trim(xmlNodePtr node, char *(*f)(char *, const char *)) {
	char *content, *trimmed;

	content = (char *) xmlNodeGetContent(node);

	trimmed = calloc(strlen(content) + 1, 1);
	f(trimmed, content);
	xmlFree(content);

	content = strdup(trimmed);
	xmlFree(trimmed);

	xmlNodeSetContent(node, BAD_CAST content);
	xmlFree(content);
}

/* Trim all text nodes in a given set of elements. */
void trim_nodes(xmlNodeSetPtr nodes, bool normalize)
{
	int i;

	for (i = 0; i < nodes->nodeNr; ++i) {
		xmlNodePtr first, last;

		/* If node has no children, no trimming is necessary. */
		if (!nodes->nodeTab[i]->children) {
			continue;
		}

		if ((first = nodes->nodeTab[i]->children)->type == XML_TEXT_NODE) {
			trim(first, strltrm);
		}

		if ((last = xmlGetLastChild(nodes->nodeTab[i]))->type == XML_TEXT_NODE) {
			trim(last, strrtrm);
		}

		if (normalize) {
			for (first = nodes->nodeTab[i]->children; first; first = first->next) {
				if (first->type == XML_TEXT_NODE) {
					trim(first, strnorm);
				}
			}
		}
	}
}

void trim_nodes_in_file(const char *path, xmlNodePtr ns, xmlNodePtr elems, bool normalize, bool overwrite)
{
	xmlDocPtr doc;
	xmlXPathContextPtr ctx;
	xmlNodePtr cur;

	doc = read_xml_doc(path);
	ctx = xmlXPathNewContext(doc);

	for (cur = ns->children; cur; cur = cur->next) {
		xmlChar *n;
		n = xmlNodeGetContent(cur);
		register_ns(ctx, (char *) n);
		xmlFree(n);
	}

	for (cur = elems->children; cur; cur = cur->next) {
		xmlChar *xpath;
		xmlXPathObjectPtr obj;
		xmlChar *e;

		e = xmlNodeGetContent(cur);

		/* If the element specifier contains a /, treat it like a
		 * literal XPath expression.
		 *
		 * Otherwise, match all elements with the same name at any
		 * position.
		 */
		if (xmlStrchr(e, '/')) {
			xpath = xmlStrdup(e);
		} else {
			xpath = xmlStrdup(BAD_CAST "//");
			xpath = xmlStrcat(xpath, e);
		}

		xmlFree(e);

		obj = xmlXPathEvalExpression(xpath, ctx);

		xmlFree(xpath);

		if (!xmlXPathNodeSetIsEmpty(obj->nodesetval)) {
			trim_nodes(obj->nodesetval, normalize);
		}

		xmlXPathFreeObject(obj);
	}

	xmlXPathFreeContext(ctx);

	if (overwrite) {
		xmlSaveFile(path, doc);
	} else {
		xmlSaveFile("-", doc);
	}

	xmlFreeDoc(doc);
}

/* Show usage message. */
void show_help(void)
{
	puts("Usage: " PROG_NAME " [-e <elem> ...] [-N <ns=URL> ...] [-fnh?] [<src>...]");
	puts("");
	puts("Options:");
	puts("  -h -?        Show usage message.");
	puts("  -e <elem>    Element to trim space on.");
	puts("  -f           Overwrite input files.");
	puts("  -N <ns=URL>  Register a namespace.");
	puts("  -n           Normalize space as well as trim.");
	puts("  --version    Show version information.");
	puts("  <src>        XML file to trim.");
	LIBXML2_PARSE_LONGOPT_HELP
}

/* Show version information. */
void show_version(void)
{
	printf("%s (xml-utils) %s\n", PROG_NAME, VERSION);
	printf("Using libxml %s\n", xmlParserVersion);
}

int main(int argc, char **argv)
{
	int i;
	xmlNodePtr ns, elems;
	bool normalize = false;
	bool overwrite = false;

	const char *sopts = "e:fN:nh?";
	struct option lopts[] = {
		{"version", no_argument, 0, 0},
		LIBXML2_PARSE_LONGOPT_DEFS
		{0, 0, 0, 0}
	};
	int loptind = 0;

	ns = xmlNewNode(NULL, BAD_CAST "namespaces");
	elems = xmlNewNode(NULL, BAD_CAST "elems");

	while ((i = getopt_long(argc, argv, sopts, lopts, &loptind)) != -1)
		switch (i) {
			case 0:
				if (strcmp(lopts[loptind].name, "version") == 0) {
					show_version();
					return 0;
				}
				LIBXML2_PARSE_LONGOPT_HANDLE(lopts, loptind);
				break;
			case 'e':
				xmlNewChild(elems, NULL, BAD_CAST "elem", BAD_CAST optarg);
				break;
			case 'f':
				overwrite = true;
				break;
			case 'N':
				xmlNewChild(ns, NULL, BAD_CAST "ns", BAD_CAST optarg);
				break;
			case 'n':
				normalize = true;
				break;
			case 'h':
			case '?':
				show_help();
				return 0;
		}

	if (optind < argc) {
		for (i = optind; i < argc; ++i) {
			trim_nodes_in_file(argv[i], ns, elems, normalize, overwrite);
		}
	} else {
		trim_nodes_in_file("-", ns, elems, normalize, overwrite);
	}

	xmlFreeNode(ns);
	xmlFreeNode(elems);

	xmlCleanupParser();

	return 0;
}
